zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -409,215 +379,372 @@ impl<F: Field> ExecutionConfig<F> {
+                            || Ok(value),
+                        )?;
+                        region.assign_advice(
+                            || "step height inv",
+                            self.num_rows_inv,
+                            offset,
+                            || Ok(value.invert().unwrap_or(F::zero())),
+                        )?;
                     }
+
+                    offset += height;
+                    last_height = height;
                 }
+                // These are still referenced (but not used) in next rows
+                region.assign_advice(
+                    || "step height",
+                    self.num_rows_until_next_step,
+                    offset,
+                    || Ok(F::zero()),
+                )?;
+                region.assign_advice(
+                    || "step height inv",
+                    self.q_step,
+                    offset,
+                    || Ok(F::zero()),
+                )?;
+
+                // If not exact:
+                // TODO: Pad leftover region to the desired capacity
+                // TODO: Enable q_step_last
+                self.q_step_last.enable(&mut region, offset - last_height)?;
+
                 Ok(())
             },
-        )?;
-
-        // TODO: Pad leftover region to the desired capacity
-        // TODO: Enable q_step_last
-
-        Ok(())
+        )
     }
 
-    /// Assign exact steps in block without padding for unit test purpose
-    pub fn assign_block_exact(
+    #[allow(clippy::too_many_arguments)]
+    fn assign_exec_step(
         &self,
-        layouter: &mut impl Layouter<F>,
+        region: &mut Region<'_, F>,
+        offset: usize,
         block: &Block<F>,
+        transaction: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+        height: usize,
+        next: Option<&(&Transaction, &ExecStep)>,
+        power_of_randomness: [F; 31],
     ) -> Result<(), Error> {
-        layouter.assign_region(
-            || "Execution step",
-            |mut region| {
-                let mut offset = 0;
-
-                self.q_step_first.enable(&mut region, offset)?;
-
-                for transaction in &block.txs {
-                    for step in &transaction.steps {
-                        let call = &transaction.calls[step.call_index];
-
-                        self.q_step.enable(&mut region, offset)?;
-                        self.assign_exec_step(&mut region, offset, block, transaction, call, step)?;
-
-                        offset += STEP_HEIGHT;
-                    }
-                }
+        // Make the region large enough for the current step and the next step.
+        // The next step's next step may also be accessed, so make the region large
+        // enough for 3 steps.
+        let region = &mut CachedRegion::<'_, '_, F>::new(
+            region,
+            power_of_randomness,
+            STEP_WIDTH,
+            MAX_STEP_HEIGHT * 3,
```

Comment:
I think it's required because I do believe it gets accessed in the witness generation for the stored expressions. The only reason it seems to work is because `MAX_STEP_HEIGHT` is currently big enough so that most of the times it also included some more rows than just needed for the two execution states that are actually being processed. I changed the height to perfectly match the height actually required for the current and next step and I got an out of bounds error (matching the height perfectly I feel doesn't make sense because it makes the logic a bit more complicated for really no reason as the region is small anyway).